pkg/service/v1alpha1/node.go

```diff
@@ -118,30 +119,48 @@ verifyPublish:
 		// The volume appears to be present in the inmomory list of volumes
 		// which implies that either the mount operation is complete
 		// or under progress.
-		// Lets verify if the mount is already completed
-		if info.Spec.Volume.DevicePath != "" {
+		if info.Spec.Volume.MountPath != mountPath {
+			// The volume appears to be mounted on a different path, which
+			// implies it is being used by some other pod on the same node.
+			// Let's wait fo the volume to be unmounted from the other path and
+			// then retry checking
+			utils.VolumesListLock.Unlock()
+			if !reVerified {
+				time.Sleep(
+					utils.VolumeWaitRetryCount * utils.VolumeWaitTimeout * time.Second,
+				)
+				reVerified = true
+				goto verifyPublish
+			}
+			return nil,
+				status.Error(
+					codes.Internal,
+					"Volume Mounted by a different pod on same node",
+				)
+			// Lets verify if the mount is already completed
+		} else if info.Spec.Volume.DevicePath != "" {
 			// Once the devicePath is set implies the volume mount has been
 			// completed, a success response can be sent back
 			utils.VolumesListLock.Unlock()
 			return &csi.NodePublishVolumeResponse{}, nil
+		} else if info.Status == apis.CSIVolumeStatusMountUnderProgress {
+			// The mount appears to be under progress lets wait for 13 seconds and
+			// reverify. 13s was decided based on the kubernetes timeout values
+			// which is 15s. Lets reply to kubernetes before it reattempts a
+			// duplicate request
+			utils.VolumesListLock.Unlock()
```

Comment:
Unlock logic is scattered across everywhere. This needs to be managed/coded better.